### ISSUE TITLE
rethinking modularization of JTR

### ISSUE THREAD

**Issue by dirkweissenborn** [uclmr/jack#140]
## Problem
I see many problems with the current modularization of JTR when trying to use it in more general settings than multiple choice with fixed numbers of candidates (which is quite a specific use-case).
Take for example span prediction. There are several issues I see (which might be due to my limited experience with JTR, please correct me if I am wrong):

1) SQuAD provides varying numbers of correct answers, it seems that this is problematic for the existing jtr format
2) it seems that the jtr_load(...) does not yet support answer spans yet
3) in general the current training script, eval hooks similar to jtr_load seem to have been written for multiple choice QA or simple classification but are hardly usable as is for other purposes
4) the model interface is quite simplistic at the moment and it is not intuitive to apply it for span prediction 
5) apart from the standard input to a QA system additional input might be expected by a model as well, e.g., something that is not easily modeled within TF like a feature for each token in the support that tells me whether that word is part of the question or not. 

## Suggestion
I suggest the following modularization which should exploit object oriented programming.

### Tasks and Models
There should be (at least) 3 kinds of tasks and corresponding types of QA models: 
* _multiple choice_: answer candidates are given apriori (this is the main application area of JTR at the moment)
    * Note: num of answers and candidates could vary -> feed extra answer partition list as input that assigns each candidate to its question
    * input: question + support + candidates (no "correct answers", this is handled by the trainer during training)
    * output: logit for each candidate, id(s) of selected candidates
    * special case - _extractive_: prediction based on given candidate spans or positions
    * special case - _open vocabulary_: predictions can anything from a large vocabulary   
* _extractive_: candidates are not provided; answers should be returned as spans (and potentially a support id when multiple are given) within the support
    * input: question + support (+ correct answers only during training, which might be needed, e.g. for predicting end given the start pointer)
    * output: logits for start positions end end positions, predicted start and end pointer
* _generative_: answers are a sequence of symbols that should be predicted by a decoder based on the given question and support
    * input: question + support (+ correct answers only during training, which is needed by the decoder)
    * output: logits sequence over token vocab for each question, sequence of token-ids

### Trainer
I propose decoupling modeling and training, which includes computation of losses, optimization etc.. Trainer should be defined for all types of models individually and be provided with gold-standard answers. They can be reused between model instances of the same type.

### Evaluator
Similar to using different trainers, evaluators (this roughly exists already as EvaluationHooks) should be implemented for the different types of models.

### Feeding input
I found it very helpful in my implementations to allow models (and trainers) themselves to enrich the input feed dicts prior to running the model. 

### Training
* all of the above should be configurable and checked whether selections fit together (e.g. prediction trainer cannot be used with extraction model)
* default options should be selected based on the QA task at hand
